sentinel/lib/lineage.js

```diff
@@ -78,9 +78,17 @@ const buildHydratedDoc = (doc, lineage) => {
     doc.contact = lineage.shift();
     current = doc.contact;
   }
-  while (current) {
-    current.parent = lineage.shift();
+  let stub = doc.parent;
+  while (true) {
+    const parent = lineage.shift() || stub;
+    if (!parent) {
+      break;
+    }
+    current.parent = parent;
     current = current.parent;
+    if (stub) {
+      stub = stub.parent;
+    }
```

Comment:
I'm confused. I feel like this code needs a comment essay. Or <grumble about types>.

What is `doc` vs. `current`? Shouldn't stub be `current.parent`? What situation would `stub` be falsey but `parent` not be? Are stubs being walked up at the same time as parents are being walked up?

Reply:
> What is doc vs. current?

Doc is whatever was passed in. Current is the parent as you walk up the tree.

> Shouldn't stub be current.parent?

I don't follow this.

> What situation would stub be falsey but parent not be?

I think that was just me being overly defensive. I'll remove the check.

> Are stubs being walked up at the same time as parents are being walked up?

Yes. I'll rename some vars and add some comments.

Reply:
So when entering the while loop `current` either equals `doc` (line 76) or it equals `doc.contact`. 

In the loop we appear to be shifting parents onto `current.parent`. Which me implies the lineage is `doc[.contact?].parent.parent.parent`

Why would the stub then start from `doc.parent`? Shouldn't it follow whatever we're shifting?

Reply:
Good spotting. When I made the readability changes a unit test started failing on exactly this. Fixed in the latest commit!